src/app/icd/ICDCheckInSender.cpp

```diff
@@ -0,0 +1,98 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDCheckInSender.h"
+
+#include <system/SystemPacketBuffer.h>
+
+#include <protocols/secure_channel/CheckinMessage.h>
+
+#include <lib/dnssd/Resolver.h>
+
+namespace chip {
+namespace app {
+
+using namespace Protocols::SecureChannel;
+
+ICDCheckInSender::ICDCheckInSender(Messaging::ExchangeManager * exchangeManager)
+{
+    VerifyOrDie(exchangeManager != nullptr);
+    mExchangeManager = exchangeManager;
+    mAddressLookupHandle.SetListener(this);
+}
+
+void ICDCheckInSender::OnNodeAddressResolved(const PeerId & peerId, const AddressResolve::ResolveResult & result)
+{
+    mResolveInProgress = false;
+
+    VerifyOrReturn(CHIP_NO_ERROR != SendCheckInMsg(mCurrentEntry, result.address),
+                   ChipLogError(AppServer, "Failed to send the ICD Check-In message"));
+}
+
+void ICDCheckInSender::OnNodeAddressResolutionFailed(const PeerId & peerId, CHIP_ERROR reason)
+{
+    mResolveInProgress = false;
+    ChipLogProgress(AppServer, "Node Address resolution failed for ICD Check-In with Node ID " ChipLogFormatX64,
+                    ChipLogValueX64(peerId.GetNodeId()));
+}
+
+CHIP_ERROR ICDCheckInSender::SendCheckInMsg(ICDMonitoringEntry & entry, const Transport::PeerAddress & addr)
+{
+
+    System::PacketBufferHandle buffer = MessagePacketBuffer::New(CheckinMessage::sMinPayloadSize);
+
+    VerifyOrReturnError(!buffer.IsNull(), CHIP_ERROR_NO_MEMORY);
+    MutableByteSpan output{ buffer->Start(), buffer->DataLength() };
+
+    // TODO retrieve Check-in counter
+    CounterType counter = 0;
+
+    ReturnErrorOnFailure(CheckinMessage::GenerateCheckinMessagePayload(entry.key, counter, ByteSpan(), output));
+
+    VerifyOrReturnError(mExchangeManager->GetSessionManager() != nullptr, CHIP_ERROR_INTERNAL);
+
+    // Using default MRP since we are not doing MRP in this context
+    Messaging::ExchangeContext * exchangeContext = mExchangeManager->NewContext(
+        mExchangeManager->GetSessionManager()->CreateUnauthenticatedSession(addr, GetDefaultMRPConfig()).Value(), nullptr);
+
+    VerifyOrReturnError(exchangeContext != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    return exchangeContext->SendMessage(MsgType::ICD_CheckIn, std::move(buffer), Messaging::SendMessageFlags::kNoAutoRequestAck);
+}
+
+CHIP_ERROR ICDCheckInSender::RequestResolve(ICDMonitoringEntry & entry, FabricTable * fabricTable)
+{
+    VerifyOrReturnError(entry.IsValid(), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(fabricTable != nullptr, CHIP_ERROR_INTERNAL);
+    const FabricInfo * fabricInfo = fabricTable->FindFabricWithIndex(entry.fabricIndex);
+    PeerId peerId(fabricInfo->GetCompressedFabricId(), entry.checkInNodeID);
+
+    AddressResolve::NodeLookupRequest request(peerId);
+    mCurrentEntry = entry;
+
+    CHIP_ERROR err = AddressResolve::Resolver::Instance().LookupNode(request, mAddressLookupHandle);
```

Comment:
Please file an issue about needing to stay active here, and have a TODO comment referencing it.

Reply:
issue created